### ISSUE TITLE
`split_when`

### ISSUE THREAD

**Issue by jferard** [erikrose/more-itertools#302]
We have three useful functions: `more_itertools.split_at(iterable, pred)`, `more_itertools.split_before(iterable, pred)` and `more_itertools.split_after(iterable, pred)` but none of them helps for a question like [this one](https://stackoverflow.com/questions/57236885/distinct-group-of-integer):

> Given an array of integers, I want to determine the number of distinct groups of integers wherein the values ascend.

To answer this question, you have to compare two consecutive values. You can play with `zip`, but its very cumbersome. A `split_when` function taking an iterable and a function of two consecutive element is easy to code, and has the following result:

    >>> myList = [1, 2, 3, 4, 3, 2, 2, 3, 1, 2, 1, 4, 2]
    >>> list(split_when(myList, lambda x, y: x > y))
    [[1, 2, 3, 4], [3], [2, 2, 3], [1, 2], [1, 4], [2]]

Obviously, you have an equivalent to `split_before` and `split_after`:

    >>> list(split_when('OneTwo', lambda _, s: s.isupper()))
    [['O', 'n', 'e'], ['T', 'w', 'o']]

    >>> list(split_when('one1two2', lambda s, _: s.isdigit()))
    [['o', 'n', 'e', '1'], ['t', 'w', 'o', '2']]

(I don't think a replacement of these functions implementations is a good idea though).

Are you interested in a PR?

**Comment by bbayles** [comment]
Can we extend the idea to work with more than two consecutive items, possibly using `windowed()`?

**Comment by jferard** [comment]
I see three drawbacks:

1) the most important: where do you split? If you have two items, you split between the two items. But if you have 3 items? After the first one? What if you need to know the values of the first two items to decide when to split? Should we add a `split_index` parameter?

2) what would the function signature be? The simple form is: 

```
    split_when(iterable : Iterable[T], f:Callable[[T, T], bool] -> Iterable[List[T]]
```

The complex form would be:

    split_when(iterable : Iterable[T], window_size: int, f:Callable[..., bool] -> Iterable[List[T]]

Or worse:

    split_when(iterable : Iterable[T], split_index: int, window_size: int, f:Callable[..., bool] -> Iterable[List[T]]


3) The implementation might become cumbersome (but `windowed` might help).

To summarize: compare to consecutive items and split (or not) between them seems natural. A more complex function would more powerful but less intuitive and maybe slow.

**Comment by bbayles** [comment]
OK, I'm sold - go ahead and make your PR with the 2-item version.

**Comment by bbayles** [comment]
Many thanks!
